File: datastructures/stack.py
class Stack:
    def __init__(self):
        self._data = []

    def push(self, item):       # O(1)
        self._data.append(item)

    def pop(self):              # O(1)
        if self.is_empty():
            raise IndexError("Stack underflow")
        return self._data.pop()

    def peek(self):             # O(1)
        if self.is_empty():
            raise IndexError("Stack is empty")
        return self._data[-1]

    def is_empty(self):
        return len(self._data) == 0

    def size(self):
        return len(self._data)

    def __repr__(self):
        return f"Stack (top -> bottom): {self._data[::-1]}"


s = Stack()
s.push(10)
s.push(20)
s.push(30)
print(s)            # Stack (top -> bottom): [30, 20, 10]
print(s.peek())     # 30
print(s.pop())      # 30
print(s.size())     # 2

File: datastructures/Stack.java
public class Stack {

    private int[] data;
    private int top;
    private int capacity;

    public Stack(int capacity) {
        this.capacity = capacity;
        this.data = new int[capacity];
        this.top = -1;
    }

    // Push - O(1)
    public void push(int item) {
        if (top == capacity - 1) {
            throw new RuntimeException("Stack overflow");
        }
        data[++top] = item;
    }

    // Pop - O(1)
    public int pop() {
        if (isEmpty()) {
            throw new RuntimeException("Stack underflow");
        }
        return data[top--];
    }

    // Peek - O(1)
    public int peek() {
        if (isEmpty()) {
            throw new RuntimeException("Stack is empty");
        }
        return data[top];
    }

    public boolean isEmpty() {
        return top == -1;
    }

    public int size() {
        return top + 1;
    }

    public void display() {
        System.out.print("Stack (top -> bottom): ");
        for (int i = top; i >= 0; i--) {
            System.out.print(data[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Stack s = new Stack(10);
        s.push(10);
        s.push(20);
        s.push(30);
        s.display();                // Stack (top -> bottom): 30 20 10
        System.out.println(s.peek());   // 30
        System.out.println(s.pop());    // 30
        System.out.println(s.size());   // 2
    }
}
